### PR TITLE
Unfold Combinator

### PR DESCRIPTION
**Pull request by alfert**
The `unfold` combinator is similar to `Stream.unfold/2` and solves #51. The combinator generates lists using an accumulator for generating the list's elements. This provides the basis for command generation used in state-machine based property testing of stateful systems, requiring to track a symbolic state during command generation.

### CODE REVIEW

File: lib/stream_data.ex [hunk 1]
```diff
@@ -823,37 +823,10 @@ defmodule StreamData do
...
-
-      data
-      |> call_n_times(next_seed, size, length, [])
-      |> LazyTree.zip()
-      |> LazyTree.map(&list_lazy_tree(&1, 0))
-      |> LazyTree.flatten()
-    end)
+  @spec list_of(t(a), keyword()) :: t([a]) when a: term()
+  def list_of(data, options \\ []) do
+    unfold(nil, fn _ -> {data, nil} end, options)
```
Comment (michalmuskala) [inline lib/stream_data.ex]: Please verify how this compares performance-wise to the previous implementation. `list_of` is used in many other combinators, so it heavily impacts overall performance of the library.

Comment (alfert) [reply lib/stream_data.ex]: Good point. Do you have a kind of testbed and performance data as bench mark from your performance optimisation activities? Would be nice to do it the same way.

Comment (alfert) [reply lib/stream_data.ex]: First results running 
`MIX_ENV=prod mix profile.fprof -e 'StreamData.list_of(StreamData.integer()) |> Enum.take(1000)'`

Test | Function | ACC(ms) | OWN(ms) | Percentage(acc)
-----|---------|-----------|-----------|---
`list_of` | `:fprof.apply_start_stop/4` | 1811.293  | 0.008 | 100.00 %
`list_of`  | anonymous `fn/3` in `StreamData.list_of/1` | 1783.694 | 8.326 | 100.00 %
`unfold`| `:fprof.apply_start_stop/4` | 2102.877 | 0.010 | 116.09 %
`unfold`|  anonymous `fn/4` in `StreamData.unfold/2` |  2073.250 | 15.602 | 116.23 %

The additional indirection of `unfold` seems to cost 16% more performance. Looks like I eat your performance optimisations. So `list_of` should not be implemented in terms of `unfold`. I will correct that.

**Comment by alfert** [comment]
I close this pull request, since it was never merged nor did the shrinking job properly.

**Pull request closed without merging**